### PR TITLE
chore: added rich push delivery tracking

### PR DESCRIPTION
**Pull request by Shahroz16**
helps: https://linear.app/customerio/issue/MOB-13/http-layer-for-nse
root ticket: https://github.com/customerio/issues/issues/11953

TLDR (work required): We need to add an HTTP layer that exists in the `Messaging Push` module when a push is received in NSE, we can just use that HTTP client to track the delivery for us. HTTP client is going to use the CIO data pipeline API endpoint. 

## Problem: 
Currently, HTTP client implementation was meant to stay in the `common` layer, but not anymore, if we are only utilizing the HTTP client rich push notification, downloading URL/ tracking metric. The implementation should be `Messaging` module-specific as well.  

## Solution:
This PR does the following:
- Add HTTP client implementation in the Messaging Push since it's the only module that utilizes it. 
- Add logic for tracking metrics for push

What this PR isn't doing:
- Clean up unused HTTP client code implementation, usage in tracking, etc. The cleaning of tracking code and unused methods of HTTP client etc would be in another PR, otherwise, they would make the review harder.

### CODE REVIEW

File: Sources/MessagingPush/MessagingPush.swift [hunk 1]
```diff
@@ -36,16 +36,31 @@ public class MessagingPush: ModuleTopLevelObject<MessagingPushInstance>, Messagi
     public static func initialize(
         configure configureHandler: ((inout MessagingPushConfigOptions) -> Void)? = nil
     ) -> MessagingPushInstance {
-        var configOptions = moduleConfig
-
         if let configureHandler = configureHandler {
-            configureHandler(&configOptions)
+            // pass current config reference to update it without needing to recreate
+            configureHandler(&moduleConfig)
```
Comment (Shahroz16) [inline Sources/MessagingPush/MessagingPush.swift:41]: `moduleConfig` is declared at the top and is atomic. Its being utilized everywhere.

File: Sources/MessagingPush/RichPush/RichPushHttpClient.swift [hunk 2]
```diff
@@ -0,0 +1,192 @@
+import CioInternalCommon
+import Foundation
+#if canImport(FoundationNetworking)
+import FoundationNetworking
+#endif
+
+// sourcery: InjectRegisterShared = "HttpClient"
+public class RichPushHttpClient: HttpClient {
```
Comment (Shahroz16) [inline Sources/MessagingPush/RichPush/RichPushHttpClient.swift:8]: Class is very similar to the `HttPClient` implementation in the `Tracking` module, but stripped down to only contain things needed and decoupled from any `Queue` logic.

File: Sources/MessagingPush/RichPush/RichPushDeliveryTracker.swift [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    let httpClient: HttpClient
+    let logger: Logger
+
+    init(httpClient: HttpClient, logger: Logger) {
+        self.httpClient = httpClient
+        self.logger = logger
+    }
+
+    func trackMetric(token: String, event: Metric, deliveryId: String, timestamp: String? = nil, onComplete: @escaping (Result<Void, HttpRequestError>) -> Void) {
+        let properties: [String: Any] = [
```
Comment (Shahroz16) [inline Sources/MessagingPush/RichPush/RichPushDeliveryTracker.swift:15]: since we don't plan on reusing this, kept as a map rather than a struct/class.

File: Sources/MessagingPush/MessagingPush.swift [hunk 4]
```diff
@@ -36,16 +36,31 @@ public class MessagingPush: ModuleTopLevelObject<MessagingPushInstance>, Messagi
...
         }
 
         shared.initializeModule()
         return shared
     }
 
+    /// MessagingPush initializer for Notification Service Extension
```
Comment (levibostian) [inline Sources/MessagingPush/MessagingPush.swift:48]: From my understanding, customers are asked to call `CustomerIO.initialize()` to provide the `writeKey` in the host app. Then, the `writeKey` can be referenced by other modules in the code. 

This change looks like we are introducing `writeKey` into the push module. This `writeKey` would only be used by the NSE for delivered push metrics. 

Assuming this is correct, could we avoid introducing `writeKey` to the push module? Having to manage multiple `writeKeys` in the SDK, depending on the context, would not be ideal. I would consider this change as introducing technical debt and I would prefer we avoid it if possible.

File: Sources/MessagingPush/RichPush/RichPushDeliveryTracker.swift [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+class RichPushDeliveryTracker {
+    let httpClient: HttpClient
+    let logger: Logger
+
+    init(httpClient: HttpClient, logger: Logger) {
+        self.httpClient = httpClient
+        self.logger = logger
+    }
+
+    func trackMetric(token: String, event: Metric, deliveryId: String, timestamp: String? = nil, onComplete: @escaping (Result<Void, HttpRequestError>) -> Void) {
```
Comment (levibostian) [inline Sources/MessagingPush/RichPush/RichPushDeliveryTracker.swift:14]:
```suggestion
    func trackMetric(token: String, deliveryId: String, timestamp: String? = nil, onComplete: @escaping (Result<Void, HttpRequestError>) -> Void) {
```

The class name and use case seems to only be for delivered tracking. I don't think that we need to have an `event` parameter?

File: Sources/MessagingPush/RichPush/RichPushDeliveryTracker.swift [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+            ],
+            "event": "Report Delivery Event"
+        ]
+
+        let endpoint: CIOApiEndpoint = .trackNSEPushDeliveryMetrics
+        guard let httpParams = HttpRequestParams(
+            endpoint: endpoint,
+            baseUrl: RichPushHttpClient.defaultAPIHost,
+            headers: nil,
+            body: try? JSONSerialization.data(withJSONObject: properties)
```
Comment (levibostian) [inline Sources/MessagingPush/RichPush/RichPushDeliveryTracker.swift:30]: Is there a reason for not using `jsonAdapter` here?

File: Sources/MessagingPush/RichPush/RichPushHttpClient.swift (hunk 2 above)
Comment (levibostian) [reply Sources/MessagingPush/RichPush/RichPushHttpClient.swift:8]: Instead of introducing a new `HttpClient` implementation inside of the push module, could we re-use the `HttpClient` inside of the Common module? 

I think that adding this file (that looks identical to the one in Common) to the push module would introduce technical debt to the project and I would like to avoid that if possible.

File: Sources/MessagingPush/MessagingPushImplementation.swift [hunk 7]
```diff
@@ -43,6 +47,14 @@ class MessagingPushImplementation: MessagingPushInstance {
         eventBusHandler.postEvent(TrackMetricEvent(deliveryID: deliveryID, event: event.rawValue, deviceToken: deviceToken))
     }
 
+    func trackMetricFromNSE(
```
Comment (levibostian) [inline Sources/MessagingPush/MessagingPushImplementation.swift:50]:
```suggestion
    func trackMetricOrFail(
```

From a code maintenance standpoint, I suggest we name functions by what they do (attempt or track metric, or fail) instead of where they are used (NSE). 

I think it could be ever better to isolate the NSE logic from the host app logic since the use case is different. 

Example: 
```swift
// Used by host app
class MessagingPushImplementation {
  trackMetric()
    eventBus...
}

// Used by NSE
class MessagingPushNSEImplementation: MessagingPushImplementation {
  override trackMetric()
    httpClient...
}
```

**Review by levibostian — CHANGES_REQUESTED** [review]
Rejecting due to concerns I have about introducing technical debt into the codebase from the PR. 

I have a feeling I am missing context behind the implementation choice. If that is true, some of these questions could be resolved by me better understanding the decisions in this PR.

File: Sources/MessagingPush/RichPush/RichPushDeliveryTracker.swift (hunk 6 above)
Comment (Shahroz16) [reply Sources/MessagingPush/RichPush/RichPushDeliveryTracker.swift:30]: I think I tried answering this in previous PRs too, so let me explain with an example. `jsonAdapter` is not that agnostic is meant to work as an adapter for json that our `tracking API` expected and accepted. 

For example in this case, if were to use the method of `jsonAdapter` 

```
 public func fromDictionary<T: Decodable>(_ dictionary: [AnyHashable: Any]) -> T? {
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: dictionary)

            return fromJson(jsonData)
        } catch {
            log.error("\(error.localizedDescription), dictionary: \(dictionary)")
        }

        return nil
    }
```

`let jsonData = try JSONSerialization.data(withJSONObject: dictionary)` it's great and solves our problem, but later you see the very next line, `fromJson(jsonData)`, it starts decoding this `json` and breaks it. 

Now I can go ahead and start creating these new methods in `jsonAdapter` but again `jsonAdapter` is biased, you see the decode being used, now that wouldn't work too because to be compatible with Segment we need to use it `iso8601` as date decoding strategy. 

```
var decoder: JSONDecoder {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .secondsSince1970
        return decoder
    }
```


Hence, I avoided using the `jsonAdapter` too aggressively or any other class for the utility that's not agnostic, for cleanup later on we can create the utility class but currently, the usage is too low to need one imho.

File: Sources/MessagingPush/RichPush/RichPushHttpClient.swift (hunk 2 above)
Comment (Shahroz16) [reply Sources/MessagingPush/RichPush/RichPushHttpClient.swift:8]: `CIOHttpClient` the implementation HttpClient in the `common` would get deleted, since its not being used anymore and was reliant on CIO api/headers etc.

Comment (levibostian) [reply Sources/MessagingPush/RichPush/RichPushHttpClient.swift:8]: I understand that the current `CIOHttpClient` is specific to the Track API. That client is going to go away since we will no longer call the Track API. 

Because the common module contains general classes that can be used by any module in the SDK, it makes the most sense to me for a general `CIOCDPHttpClient` to be contained inside of the common module.

File: Sources/MessagingPush/RichPush/RichPushDeliveryTracker.swift (hunk 6 above)
Comment (levibostian) [reply Sources/MessagingPush/RichPush/RichPushDeliveryTracker.swift:30]: If `JsonAdapter` is not being used in multiple places, then I am OK with not making a change in this current PR. 


The implementation of `JsonAdapter` today could be bias towards Track.

File: Sources/MessagingPush/MessagingPushImplementation.swift (hunk 7 above)
Comment (levibostian) [reply Sources/MessagingPush/MessagingPushImplementation.swift:50]: Although I don't like having 2 `trackMetrics` functions with different bodies, we know we have more refactors coming for the NSE target. Therefore, I am OK with holding off on change to then.

**Review by levibostian — APPROVED** [review]

File: Sources/MessagingPush/RichPush/RichPushHttpClient.swift (hunk 2 above)
Comment (Shahroz16) [reply Sources/MessagingPush/RichPush/RichPushHttpClient.swift:8]: `HttpClient` lives in common, if we were to add `CIOCDPHttpClient` in the common as well, we are going to make a decision to bind `common` module with `datapipeline` which doesn't make much sense, since we have multiple modules `push/in-app` and more in future that might just be feature modules independent of datapipeline. 

For example, Gist is using only its own http client for gist services, it doesn't want to utilize `CIOCDPHttpClient`. 

Additionally, `CIOCDPHttpClient` requires, `writeKey` and we won't have that in `common` module.
